### PR TITLE
feature: tests

### PR DESCRIPTION
**Pull request by Kupov**

### CODE REVIEW

File: test/SaveControl/SaveControlSpec.js [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+    it('action TABLE_EDITOR_LOAD_SUCCESS', function() {
+      expect(reducer(
+        deepFreeze(state),
+        {
+          type: 'TABLE_EDITOR_LOAD_SUCCESS',
+          payload: {
+            rows: previousState
+          }
+        }
+      )).toEqual({...state, prevState: previousState});
```
Comment (apress-hound) [inline test/SaveControl/SaveControlSpec.js]: Expected '}' to match '{' from line 301 and instead saw '...'.<br>Expected ')' and instead saw 'state'.<br>Missing semicolon.<br>Expected '}' to match '{' from line 292 and instead saw ':'.<br>Expected ')' and instead saw 'previousState'.<br>Unrecoverable syntax error. (65% scanned).

File: test/SaveControl/SaveControlSpec.js [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+      prevState: [],
+      isProgress: false,
+      fetchDiff: false,
+      waitingState: [],
+      saveState: []
+    };
+
+    it('action none', function() {
+      expect(reducer(deepFreeze(state), {type: null, payload: {}}))
+        .toEqual({...state});
```
Comment (apress-hound) [inline test/SaveControl/SaveControlSpec.js]: Expected '}' to match '{' from line 289 and instead saw '...'.<br>Expected ')' and instead saw 'state'.<br>Missing semicolon.

File: test/SaveControl/SaveControlSpec.js [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+    const sagaSaveCreateDiff = sagas.saveCreateDiff({
+      payload: {
+        curState: currentState,
+        prevState: previousState
+      }
+    });
+    const sagasSave = sagas.save();
+    let next = {};
+
+    const getSave = () => ({
```
Comment (apress-hound) [inline test/SaveControl/SaveControlSpec.js]: 'getSave' is defined but never used.

File: test/SaveControl/SaveControlSpec.js [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+      expect(next.value).toEqual(call(sagas.pollingJob, '123'));
+    });
+
+    it('sagas save step 6', function() {
+      next = sagasSave.next({succeeded: true, payload: {}});
+      expect(next.value).toEqual(put({
+        type: 'TABLE_EDITOR_ROW_ADD_DEFAULT_ID',
+        payload: [{
+            id: -7,
+            record_id: -100000
```
Comment (apress-hound) [inline test/SaveControl/SaveControlSpec.js]: Identifier 'record_id' is not in camel case.

File: test/SaveControl/SaveControlSpec.js [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+      expect(next.value).toEqual(select(sagas.getSave));
+    });
+
+    it('sagas save step 3', function() {
+      next = sagasSave.next({saveState: mergeDifference});
+      expect(next.value).toEqual(call(sagas.putSave, mergeDifference));
+    });
+
+    it('sagas save step 4', function() {
+      next = sagasSave.next({job_id: '123'});
```
Comment (apress-hound) [inline test/SaveControl/SaveControlSpec.js]: Identifier 'job_id' is not in camel case.

File: test/SaveControl/SaveControlSpec.js [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+      {
+        check: {
+          common: {
+            id: -7
+          }
+        },
+        destroy: true
+      }
+    ]
+    let next = {};
```
Comment (apress-hound) [inline test/SaveControl/SaveControlSpec.js]: 'next' is defined but never used.

File: test/SaveControl/SaveControlSpec.js [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+
+describe('sagas save control', function() {
+  describe('difference', function() {
+    const sagaSaveCreateDiff = sagas.saveCreateDiff({
+      payload: {
+        curState: currentState,
+        prevState: previousState
+      }
+    });
+    const getSave = () => ({waitingState});
```
Comment (apress-hound) [inline test/SaveControl/SaveControlSpec.js]: 'getSave' is defined but never used.

File: test/SaveControl/SaveControlSpec.js [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+
+window.app = {
+  config: {
+    urlSaveTiger: '/table/save',
+  }
+};
+
+describe('sagas save control', function() {
+  describe('difference', function() {
+    const sagaSaveCreateDiff = sagas.saveCreateDiff({
```
Comment (apress-hound) [inline test/SaveControl/SaveControlSpec.js]: 'sagaSaveCreateDiff' is defined but never used.

File: test/SaveControl/SaveControlSpec.js [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+        }
+      });
+    });
+
+    for (var i = 0; i < 3; i++) {
+      (function(i) {
+        it('get difference row', function() {
+          expect(sagas.getRowDifference(currentState[i], previousState[i])).toEqual(differenceState[i]);
+        });
+      })(i)
```
Comment (apress-hound) [inline test/SaveControl/SaveControlSpec.js]: Missing semicolon.

File: test/SaveControl/SaveControlSpec.js [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+          }
+        }
+      });
+    });
+
+    it('get difference product group cell', function() {
+      expect(sagas.getProductGroupDifference(currentState[2], previousState[2])).toEqual({
+        product_group: {
+          common: {
+            parent_id: 2
```
Comment (apress-hound) [inline test/SaveControl/SaveControlSpec.js]: Identifier 'parent_id' is not in camel case.

**Comment by Kupov** [comment]
@j-u-p-iter

**Comment by dontuse** [comment]
это зачем ?

File: test/SaveControl/SaveControlSpec.js [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+import previousState from './previousState.json';
+import differenceState from './differenceState.json';
+import differenceStateTransform from './differenceStateTransform.json';
+import waitingState from './waitingState.json';
+import mergeDifference from './mergeDifference.json';
+import currentStateMergeInvalidDifference from './currentStateMergeInvalidDifference.json';
+import * as sagas from '../../src/SaveControl/sagas.js';
+import * as actions from '../../src/SaveControl/actions.js';
+import reducer from '../../src/SaveControl/reducer.js';
+import _isEqual from 'lodash/isEqual';
```
Comment (dontuse) [inline test/SaveControl/SaveControlSpec.js]: это тут точно не нужно.

**Pull request closed without merging**